### PR TITLE
FF-2998: Postgres primitive array support.

### PR DESCRIPTION
**Pull request by jcustenborder**
## Problem

Postgres wants arrays.

## Solution

Give Postgres arrays.

##### Does this solution apply anywhere else?
- [ ] yes
- [X] no

##### Testing done:
- [X] Unit tests
- [X] Integration tests
- [ ] System tests
- [ ] Manual tests

### CODE REVIEW

File: src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java [hunk 1]
```diff
@@ -366,6 +372,74 @@ protected void formatColumnValue(
...
+      Object value
+  ) throws SQLException {
+
+    switch (schema.type()) {
+      case ARRAY: {
+        Class<?> valueClass = value.getClass();
+        Object newValue = null;
+        if (valueClass.isArray()) {
+          newValue = value;
+        } else {
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java:390]: Maybe check that valueClass is a Collection here, and have another else that throws if we try to bind a non-array, non-collection type to an array?

File: src/test/java/io/confluent/connect/jdbc/sink/integration/PostgresDatatypeIT.java [hunk 2]
```diff
@@ -213,6 +216,118 @@ public void testWriteToTableWithUuidColumn() throws Exception {
...
+          assertJDBCArray(rs, "friendnames", struct);
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testWriteToTableWithIntArrayColumnMissingFields() throws SQLException, InterruptedException {
+    createTableWithIntArrayColumnsMissing();
+    props.put(JdbcSinkConfig.AUTO_CREATE, "true");
```
Comment (ncliang) [inline src/test/java/io/confluent/connect/jdbc/sink/integration/PostgresDatatypeIT.java:257]: auto_create is not required for this testcase. testAutoCreateTableWithPrimitiveTypeArrayColumn should be its own testcase.

**Review by ncliang — COMMENTED** [review]
Thanks @jcustenborder . I see that you are building on https://github.com/confluentinc/kafka-connect-jdbc/pull/805 and addressing comments on that PR. I have some more minor comments here. High level, how much effort do you think it would be to add this support in the generic dialect? Is it mostly testing to figure out which driver implementations do not support binding primitive arrays with `bindObject` and array type? We have some IT that are testing with MySql and MSSQL dialects, and tests using embedded Derby. If we just covered these dialects with some manual testing for others, I think that could give us enough confidence to move it to the generic dialect. This does not need to be part of this PR, but could be covered in a followup PR as well.

File: src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java (hunk 1 above)
Comment (jcustenborder) [reply src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java:394]: Unified array and list codepaths.

File: src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java [hunk 3]
```diff
@@ -366,6 +374,80 @@ protected void formatColumnValue(
...
+          case FLOAT32:
+            newValue = valueCollection.toArray(new Float[0]);
+            break;
+          case INT64:
+            newValue = valueCollection.toArray(new Long[0]);
+            break;
+          default:
+            break;
+        }
+
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java:437]:
```suggestion
```

**Review by ncliang — APPROVED** [review]
Thanks @Faqa , @jcustenborder , the change LGTM. We can expand to support arrays in other dialects in another PR.

**Comment by jcustenborder** [comment]
Thanks buddy!

**Comment by aakashnshah** [comment]
@Iskuskov I'm not sure if we have timelines for this, usually we do these releases when a new version is warranted (typically when a feature is merged to master like this). However, I don't have enough context on this to know if it is a feature that needs to be immediately released. I tagged the people above that seem to be most knowledgable about this, hopefully they can give a better answer.

**Comment by Iskuskov** [comment]
@jcustenborder, hi! thank you very much for the great feature once again! Is it correct this PR supports arrays only on sink side, not on source?

**Comment by NathanNam** [comment]
This feature should be available in 10.2.0 or above.

**Pull request merged**